Add new blog for password authentication (#317)
prestodb/prestodb.github.io: website/blog/2026-04-29-presto-password-authentication.md

---
author: Pratyaksh Sharma
authorURL: https://www.linkedin.com/in/pratyaksh-sharma-60593769/
title: Password Authentication setup on Local
---

> Authentication and authorization are two main pillars for data security. While a Presto cluster can be setup to run 

Reviewer: ```suggestion
Authentication and authorization are two main pillars for data security. While a Presto cluster can be set up to run 
```

2. How are TLS connections different from mTLS connections?
3. What is a keystore and a truststore?
4. What is the role of keystore and truststore in SSL connections?
5. Basic Java tools like `keytool` for managing keystores, truststores and certificates.

This article will introduce these topics before we begin setting up the Presto server.

## SSL Handshakes

Let us start with some theory that I feel is important to understand before going to hands on. 
SSL handshakes are a mechanism by which client and server establish the trust and logistics required to secure their connection 
over the network. SSL handshake involves many steps, but we go over only the ones relevant for this blog: 

 - Client provides a list of possible SSL versions and cipher suites to use.
 - Server agrees on a particular SSL version and cipher suite, responding with its certificate from its keystore.
 - Client verifies the authenticity of this certificate by checking it against a list of trusted Certificate Authorities (CAs) using its truststore.
 
### Keystore and Truststore

Keystore and Truststore are used to manage the keys and certificates required for secure communication. 

> The Keystore stores your private keys and the corresponding public certificates. During a SSL handshake, the Presto server

Reviewer: ```suggestion
* The Keystore stores your private keys and the corresponding public certificates. During a SSL handshake, the Presto server
```

uses its keystore to present its identity to the client such as Presto CLI. 
* A truststore stores only public certificates, typically from trusted Certificate Authorities. Clients such as Presto CLI
will make use of its truststore to verify the certificate presented by the Presto coordinator. If the server's certificate or

> the corresponding CA is not in the client's truststore, connection fails with `SSLHandshakeException`.

Reviewer: ```suggestion
the corresponding CA is not in the client's truststore, the connection fails with `SSLHandshakeException`.
```

It is important to note that Java provides default truststore (`cacerts`) but there is no default keystore. So during a handshake, 
the connection can succeed when the server presents a certificate which is issued by well known CAs because the certificates issued by these CAs are 

> already present in the default truststore. In the case of local testing, the certificate could be a self-signed one, and 

Reviewer: ```suggestion
already present in the default truststore. When testing, the certificate could be a self-signed one, and 
```

a custom truststore will be needed which contains the presented certificate.

### TLS and mTLS

mTLS stands for mutual TLS (Transport Layer Security). The key difference between TLS and mTLS is that TLS uses a one way handshake where only the
server presents its certificate to the client and the client verifies that certificate using its truststore. However, mTLS
adds a mandatory second step where the client also presents its certificate to the server, allowing the server to authenticate 
the client, ensuring a strict two-way, zero trust security. 
In a TLS handshake, the client only needs a truststore to verify the server's identity, while in a mTLS handshake, both server and client 

> need corresponding keystore and truststore to present the certificate and verify the other's identity respectively. 

Reviewer: ```suggestion
need a keystore and truststore to present the certificate and verify the other's identity. 
```
